subcerts.go

```diff
@@ -141,67 +148,71 @@ func (cred *credential) marshal() ([]byte, error) {
 	}
 
 	// Write the length of the public_key field.
-	binary.BigEndian.PutUint16(serialized[4:], uint16(len(serializedPublicKey)))
+	binary.BigEndian.PutUint16(serialized[paramsLen:], uint16(len(serializedPublicKey)))
 
-	// Write the public key.
-	return append(serialized, serializedPublicKey...), nil
+	// Write the public_key field.
+	serialized = append(serialized, serializedPublicKey...)
+	cred.raw = serialized
+	return serialized, nil
 }
 
 // unmarshalCredential decodes a credential and returns it.
 func unmarshalCredential(serialized []byte) (*credential, error) {
-	// Bytes 0-3 are the validity time field; bytes 4-6 are the length of the
-	// serialized SubjectPublicKeyInfo.
-	if len(serialized) < 6 {
+	paramsLen := 8
+
+	// Bytes 0:4 are the valid_time field, 4:6 are the scheme field, 6:8 are the
+	// version field, and 8:10 are the length of the serialized
+	// SubjectPublicKeyInfo.
+	if len(serialized) < paramsLen+2 {
 		return nil, errors.New("credential is too short")
 	}
 
-	// Parse the validity time.
+	// Parse the valid_time, scheme, and version fields.
 	validTime := time.Duration(binary.BigEndian.Uint32(serialized)) * time.Second
+	scheme := SignatureScheme(binary.BigEndian.Uint16(serialized[4:]))
+	version := binary.BigEndian.Uint16(serialized[6:])
 
 	// Parse the SubjectPublicKeyInfo.
-	pk, scheme, err := unmarshalSubjectPublicKeyInfo(serialized[6:])
+	pk, err := unmarshalSubjectPublicKeyInfo(serialized[paramsLen+2:])
 	if err != nil {
 		return nil, err
 	}
 
-	return &credential{validTime, pk, scheme}, nil
+	return &credential{
+		raw:                         serialized,
+		validTime:                   validTime,
+		expectedCertVerifyAlgorithm: scheme,
+		expectedVersion:             version,
+		publicKey:                   pk,
+	}, nil
 }
 
 // unmarshalSubjectPublicKeyInfo parses a DER encoded SubjectPublicKeyInfo
-// structure into a public key and its corresponding algorithm.
-func unmarshalSubjectPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+// structure into a public key.
+func unmarshalSubjectPublicKeyInfo(serialized []byte) (crypto.PublicKey, error) {
 	publicKey, err := x509.ParsePKIXPublicKey(serialized)
 	if err != nil {
-		return nil, 0, err
+		return nil, err
 	}
 
 	switch pk := publicKey.(type) {
```

Comment:
Can you remove this code bloat? I would prefer something like:
```
if publicKey.(type) == *ecdsa.PublicKey {
return publicKey.(type), nil
}

return nil, fmt.Errorf()
```
Also this function is not needed anymore. Also to be removed

Reply:
What you suggest isn't syntactically correct, because `.(type)` can only happen in a `switch`:
```
GOROOT/linux_amd64/src/crypto/tls/subcerts.go:199:7: use of .(type) outside type switch
GOROOT/linux_amd64/src/crypto/tls/subcerts.go:199:18: type *ecdsa.PublicKey is not an expression
```
But I'm pretty sure this could be done with the `reflect` package. Want me to do that?

> Also this function is not needed anymore. Also to be removed
Which function?

Reply:
nop, google "type assertions in go". hint: we are using it a lot in tris